doc/subsystems/shell/shell.rst

```diff
@@ -0,0 +1,294 @@
+
+For API documentation please refer :file:`include/shell/shell.h`
+
+Usage
+*****
+
+Use the :c:macro:`SHELL_DEFINE` macro to create an instance of the shell.
+It is important to pass correct newline character to this macro, otherwise
+the shell will not respond correctly to the Enter button.
+
+The following code shows a simple use case of this library:
+
+.. code-block:: c
+
+	/* Defining shell backend */
+	SHELL_UART_DEFINE(shell_transport_uart);
+
+	/* Creating shell instance */
+	SHELL_DEFINE(uart_shell, "uart:~$ ", &shell_transport_uart, '\r', 10);
+
+	void main(void)
+	{
+		(void)shell_init(&uart_shell, NULL, true, true, LOG_LEVEL_INF);
+	}
+
+	/* Creating subcommands (level 1 command) array for command "demo".
+	 * Subcommands must be added in alphabetical order
+	 */
+	SHELL_CREATE_STATIC_SUBCMD_SET(sub_demo)
+	{
+		/* Alphabetically sorted. */
+		SHELL_CMD(params, NULL, "Print params command.",
+						       cmd_demo_params),
+		SHELL_CMD(ping,   NULL, "Ping command.", cmd_demo_ping),
+		SHELL_SUBCMD_SET_END /* Array terminated. */
+	};
+	/* Creating root (level 0) command "demo" */
+	SHELL_CMD_REGISTER(demo, &sub_demo, "Demo commands", NULL);
+
+	/* Creating root (level 0) command "version" */
+	SHELL_CMD_REGISTER(version, NULL, "Show kernel version", cmd_version);
+
+
+Users may use the Tab key to complete a command/subcommand or to see the
+available subcommands for the currently entered command level.
+For example, when the cursor is positioned at the beginning of the command
+line and the Tab key is pressed, the user will see all root (level 0)
+commands:
+
+.. code-block:: none
+
+	  clear  demo  shell  history  log  resize  version
+
+
+.. note::
+	To view the subcommands that are available for a specific command, you must first type a space after this command and then hit Tab.
+
+These commands are registered by various modules, for example:
+
+* ``clear``, ``shell``, ``history``, and ``resize`` are built-in commands which have been registered by :file:`subsys/shell/.shell.c` ,
+* ``demo`` and ``version`` have been registered in example code above by main.c,
+* log has been registered by :file:`subsys/logging/log_cmds.c`
+
+Then, if a user types a ``demo`` command and presses the Tab key, the shell
+will only print the subcommands registered for this command:
+
+.. code-block:: none
+
+	  params  ping
+
+
+Commands
+********
+
+shell commands are organized in a tree structure and grouped into the following
+types:
+
+* Root command (level 0): Kept in the section variable area.
+* Static subcommand (level > 0): Number and syntax must be known during compile time. Kept in the software module.
+* Dynamic subcommand (level > 0): Number and syntax does not need to be known during compile time. Created in the software module.
+ 
+Creating commands
+=================
+
+Use the following macros for adding shell commands:
+
+* :c:macro:`SHELL_CMD_REGISTER` - Create root command. Each and every root command shall have unique syntax.
```

Comment:
Saying all root commands have "unique syntax" doesn't sound right.  Actually their syntax is similar if not the same:  command name followed by arguments.  I think you mean all the root commands must be unique, as in I can't have two commands with the same name.